hooks/tk-multi-publish2/basic/publish_session.py

```diff
@@ -0,0 +1,712 @@
+            - checked: If True, the plugin will be checked in the UI, otherwise
+                it will be unchecked. Optional, True by default.
+
+        :param settings: Dictionary of Settings. The keys are strings, matching
+            the keys returned in the settings property. The values are `Setting`
+            instances.
+        :param item: Item to process
+
+        :returns: dictionary with boolean keys accepted, required and enabled
+        """
+
+        path = _session_path()
+
+        if not path:
+            # the session has not been saved before (no path determined).
+            # provide a save button. the session will need to be saved before
+            # validation will succeed.
+            self.logger.warn(
+                "The Maya session has not been saved.",
+                extra=_get_save_as_action()
+            )
+
+        self.logger.info(
+            "Maya '%s' plugin accepted the current Maya session." %
+            (self.name,)
+        )
+        return {
+            "accepted": True,
+            "checked": True
+        }
+
+    def validate(self, settings, item):
+        """
+        Validates the given item to check that it is ok to publish. Returns a
+        boolean to indicate validity.
+
+        :param settings: Dictionary of Settings. The keys are strings, matching
+            the keys returned in the settings property. The values are `Setting`
+            instances.
+        :param item: Item to process
+        :returns: True if item is valid, False otherwise.
+        """
+
+        publisher = self.parent
+        path = _session_path()
+
+        if not path:
+            # the session still requires saving. provide a save button.
+            # validation fails.
+            self.logger.error(
+                "The Maya session has not been saved.",
+                extra=_get_save_as_action()
+            )
+            return False
+
+        # ensure we have an updated project root
+        project_root = cmds.workspace(q=True, rootDirectory=True)
+        item.properties["project_root"] = project_root
+
+        # warn if no project root could be determined.
+        if not project_root:
+            self.logger.warning(
+                "Your session is not part of a maya project.",
+                extra={
+                    "action_button": {
+                        "label": "Set Project",
+                        "tooltip": "Set the maya project",
+                        "callback": lambda: mel.eval('setProject ""')
+                    }
+                }
+            )
+
+        # get the path in a normalized state. no trailing separator,
+        # separators are appropriate for current os, no double separators,
+        # etc.
+        path = sgtk.util.ShotgunPath.normalize(path)
+
+        # determine the publish path, version, type, and name
+        publish_info = self._get_publish_info(path, settings)
```

Comment:
The logic for calculating publish path, type, name, etc is moved into this method which accounts for ZC and templates.

Reply:
Does this mean the work area path is not templated and that we only use templates for outputs? How does this work in conjunction with workfiles?

Reply:
This is something we should discuss. I've only touched the session publish hook so far. It is now setup to use the work file template to match against the current session path. I need to update the "begin version control" hook to also default to a work file template when saving the file. I did update the "save as" method used by the publish hook to use wf2 when prompting for a file name, so it's part way there. My biggest concern is the configuration and the fact that now we need to configure templates per plugin. Anyway, let's discuss.